File: 프로그래머스/2/68936. 쿼드압축 후 개수 세기/쿼드압축 후 개수 세기.cpp
#include <string>
#include <vector>

using namespace std;

vector<int> cnt(2);

bool same(int sz, int r, int c, const vector<vector<int>> &arr) {
    int base = arr[r][c];
    for (int i = r; i < r + sz; i++) {
        for (int j = c; j < c + sz; j++) {
            if (arr[i][j] != base) return false;
        }
    }
    cnt[base]++;
    return true;
}

void dfs(int sz, int r, int c, const vector<vector<int>> &arr) {
    if (same(sz, r, c, arr)) return;
    sz /= 2;
    dfs(sz, r, c, arr);
    dfs(sz, r + sz, c, arr);
    dfs(sz, r, c + sz, arr);
    dfs(sz, r + sz, c + sz, arr);
}

vector<int> solution(vector<vector<int>> arr) {
    dfs(arr.size(), 0, 0, arr);
    return cnt;
}

File: 프로그래머스/2/68936. 쿼드압축 후 개수 세기/쿼드압축 후 개수 세기.kt
class Solution {
    val counts = intArrayOf(0, 0)

    fun solution(arr: Array<IntArray>): IntArray {
        arr.zip(0, 0, arr.size)
        return counts
    }

    fun Array<IntArray>.zip(row: Int, column: Int, length: Int){
        val firstElement = this[row][column]

        // 모두 같은 값인지 확인
        var areSameValue = true
        val columnRange = (column) until (column + length)
        var arr: IntArray

        for(r in (row) until (row + length)){
            arr = this[r]

            for(c in columnRange){
                areSameValue = arr[c] == firstElement
                if(!areSameValue){
                    break
                }
            }

            if(!areSameValue){
                break
            }
        }

        // 0, 1만으로 이루어진 경우 0 또는 1의 개수를 증가시키고 리턴
        if(areSameValue){
            counts[firstElement]++
            return
        }

        val halfLength = length / 2

        val centerRow = row + halfLength
        val centerColumn = column + halfLength

        val loopRanges = arrayOf(
            row to column,
            row to centerColumn,
            centerRow to column,
            centerRow to centerColumn
        )

        // 4분할 한다
        for((startRow, startColumn) in loopRanges){
            zip(startRow, startColumn, halfLength)
        }
    }
}